### DISCUSSION TITLE
Selfhosting

### DISCUSSION THREAD

**Discussion by eliabieri** [shuttle-hq/shuttle#976] [Q&A]
Can I self-host `shuttle`?
I did not find any documentation on this.

**Comment by wvffle** [comment]
I'm also interested in self-hosting / self-deployment. Unfortunately it seems like it was a possibility in the past but curerntly there isn't any. 

According to the discussion on https://github.com/shuttle-hq/shuttle/issues/81, there has been a Terrarofm module that isn't publicly available anymore due to some security checks.

The website states that you can self host but any information on how to do this is hard to find.
![image](https://github.com/shuttle-hq/shuttle/assets/13330620/324e599e-98a5-4d21-b40a-05167c1af5a3)

@chesedo, @christos-h : could you please clarify shuttle's state on self deployments? Are the the security checks mentioned in the linked issue already done?

If you aren't willing to provide the info on how to self deploy, or you've decided to close source the server code (I don't really know if it's in another repo or not since any information on that topic is hard to find), then please update the outdated information on your website.

**Comment by MGlolenstine** [comment]
I'm also curious about self-hosting and I'm wondering what the current state of it is.

The website still states the abovementioned.

**Reply by wvffle** [reply]
@MGlolenstine I doubt that anyone from shuttle's team would ever respond to this discussion. Maybe it would be a good idea to ask them on their discord or twitter? I don't have an account on any social media websites, so I cannot do it really.

**Reply by MGlolenstine** [reply]
Thanks for the heads-up!

After reading a bit through the issue you've tagged above, I found [this](https://github.com/shuttle-hq/shuttle/issues/81#issuecomment-1827672595) comment 2 weeks ago.
I guess that while it is on the roadmap, it might not be happening for a while.

**Reply by wvffle** [reply]
Yeah, I've just revised the linked issue and found the same comment. Judging by the first and second FAQ entry, it seems like you're able to run the project locally without any problem.

**Comment by wvffle** [comment]
@oddgrd, in https://github.com/shuttle-hq/shuttle/issues/81#issuecomment-1827672595 you've pointed us to the last FAQ entry.

Could you explain the second point of that entry?

> Running the Shuttle stack on your own server. A similar setup which can give you an experience more similar to the full Shuttle experience.

What does it mean? Does it mean to simply run the `cargo shuttle run --release` on my own server and call it a day? 

Is there an option to build a single executable file that I could run on the server?

**Reply by MGlolenstine** [reply]
I've understood it (2nd option) as an alternative that could still be deployed to, just with less features than the whole `shuttle.rs` stack, which is what the option `3` stands for.

**Comment by fatfingers23** [comment]
I run a self-hosted Shuttle application a API and Discord bot.  I just run these inside of a docker container and run with `cargo shuttle run --release` and that has worked well for me. Def the easiest way to get going, can use my [docker-compose.yml](https://github.com/fatfingers23/trackscape-discord-bot/blob/main/docker-compose.yml) and [Dockerfile](https://github.com/fatfingers23/trackscape-discord-bot/blob/main/Dockerfile) as a template. This is what I recommend doing for self hosted, it's by far the simplest one.

The next way you can do it, that I know, is you can selfhost the Shuttle stack on your own server. The best way to get started with this would probably be following  [DEVELOPING.MD](https://github.com/shuttle-hq/shuttle/blob/main/DEVELOPING.md). This will help you get your local version going and you can even use your installed version of cargo shuttle with ` --api-ur http://your-hosting-url` (You will probably see some breaking changes as shuttle stack gets updated if you don't keep it updated).

**Reply by wvffle** [reply]
You've done some great work with this Dockerfile. It basically answers all of my questions.

**Reply by oddgrd** [reply]
Thanks for sharing @fatfingers23! This is indeed the recommended way to do this for now.
